File: Tahap1/LELA.py
import cv2
import numpy as np
import time
cap = cv2.VideoCapture(0)

def get_contours(hsv, lower, upper):
    mask = cv2.inRange(hsv, lower, upper)
    contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)    
    return contours
    
 
    
recording = False
start_time = None
video_writer = None
frame_size = None
fps = 30.0 

while True:
    ret, frame = cap.read()
    frame = cv2.resize(frame, (360, 240))
    foto=frame
    
    if not ret:
        print("Gagal membaca frame dari kamera.")
        break

    hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)
    lower_red = np.array([5, 100, 165])
    upper_red = np.array([21, 255, 255])
    contours = get_contours(hsv, lower_red, upper_red)

    if contours:
        largest_contour = max(contours, key=cv2.contourArea)
         
        x, y, w, h = cv2.boundingRect(largest_contour)
        cv2.rectangle(frame, (x, y), (x + w, y + h), (0, 255, 0), 2)  # Gambar kotak
        cv2.putText(frame, 'Hotspot', (x, y - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.9, (0, 255, 0), 2)  # Tampilkan teks
        if not recording:
            
            recording = True
            start_time = time.time()
            frame_size = (frame.shape[1], frame.shape[0])
            fourcc = cv2.VideoWriter_fourcc(*'mp4v')  # Codec video (mp4v untuk MP4)
            video_writer = cv2.VideoWriter(f'recording_{time.time()}.mp4', fourcc, fps, frame_size)
            
    if recording and time.time() - start_time >= 60:
        recording = False
        video_writer.release()
        
    if recording:
        video_writer.write(frame)

    cv2.imshow('Pengenalan Hotspot', frame)
    
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break
cap.release()
cv2.destroyAllWindows()

File: Tahap1/LELA.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <chrono>

using namespace cv;
using namespace std;
using namespace std::chrono;

vector<vector<Point>> getContours(Mat hsv, Scalar lower, Scalar upper) {
    Mat mask;
    inRange(hsv, lower, upper, mask);
    vector<vector<Point>> contours;
    findContours(mask, contours, RETR_EXTERNAL, CHAIN_APPROX_SIMPLE);
    return contours;
}

int main() {
    VideoCapture cap(0);
    if (!cap.isOpened()) {
        cout << "Gagal membuka kamera." << endl;
        return -1;
    }

    bool recording = false;
    high_resolution_clock::time_point start_time;
    VideoWriter video_writer;
    Size frame_size(640, 480);
    double fps = 30.0;

    while (true) {
        Mat frame;
        cap >> frame;
        resize(frame, frame, frame_size);

        if (frame.empty()) {
            cout << "Gagal membaca frame dari kamera." << endl;
            break;
        }

        Mat hsv;
        cvtColor(frame, hsv, COLOR_BGR2HSV);
        Scalar lower_red(5, 100, 165);////////////HSV
        Scalar upper_red(21, 255, 255);///////////HSV
        vector<vector<Point>> contours = getContours(hsv, lower_red, upper_red);

        if (!contours.empty()) {
            vector<Point> largest_contour = *max_element(contours.begin(), contours.end(),
                [](const vector<Point>& a, const vector<Point>& b) {
                    return contourArea(a) < contourArea(b);
                });

            Rect bounding_rect = boundingRect(largest_contour);
            rectangle(frame, bounding_rect, Scalar(0, 255, 0), 2);
            putText(frame, "Hotspot", Point(bounding_rect.x, bounding_rect.y - 10),
                    FONT_HERSHEY_SIMPLEX, 0.9, Scalar(0, 255, 0), 2);

            if (!recording) {
                recording = true;
                start_time = high_resolution_clock::now();
                string filename = "recording_" + to_string(duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count()) + ".mp4";
                int fourcc = VideoWriter::fourcc('m', 'p', '4', 'v');
                video_writer.open(filename, fourcc, fps, frame_size);
            }
        }

        if (recording) {
            auto elapsed_time = duration_cast<seconds>(high_resolution_clock::now() - start_time).count();
            if (elapsed_time >= 60) {
                recording = false;
                video_writer.release();
            } else {
                video_writer.write(frame);
            }
        }

        imshow("Pengenalan Hotspot", frame);

        if (waitKey(1) == 'q') {
            break;
        }
    }

    cap.release();
    destroyAllWindows();

    return 0;
}
